### PR TITLE
Clear Ids for killed Pods in batches in Kubelet#podKiller

### LINKED ISSUE #77168: Clear Ids for killed Pods in batches in Kubelet#podKiller
**What would you like to be added**:
Change killing Set to map which holds Ids of Pods which are killed, along with the timestamp killing starts.
On periodic basis, obtain lock, remove Pod Ids whose killing start time is sufficiently old from the killing map.

**Why is this needed**:
Currently we clear Pod Id as soon as the Pod is killed where lock acquisition is needed.
This incurs extra locking overhead.

### PR DESCRIPTION
**Pull request by tedyu**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Currently we clear Pod Id as soon as the Pod is killed where lock acquisition is needed.
This incurs extra locking overhead.

Change killing Set to map which holds Ids of Pods which are killed, along with the timestamp killing starts.
On periodic basis, obtain lock, remove Pod Ids whose killing start time is sufficiently old from the killing map.

**Which issue(s) this PR fixes**:
Fixes #77168

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -1092,11 +1094,19 @@ func (kl *Kubelet) podKiller() {
 				if err != nil {
 					klog.Errorf("Failed killing the pod %q: %v", runningPod.Name, err)
 				}
-				lock.Lock()
-				killing.Delete(string(runningPod.ID))
-				lock.Unlock()
 			}(apiPod, runningPod)
 		}
+		// clean every 10 minutes
+		if time.Now().Unix()-timeLastClean > 600 {
```
Comment (mattjmcnaughton) [inline pkg/kubelet/kubelet_pods.go]: Are pod Ids guaranteed to be unique (i.e. is it impossible that two pods would have the same id, lets say `abc`, in the same 10 minute window)? 

For example, imagine we delete pod `abc` at 10:00 am. Then at 10:01, we create another pod which also has id `abc` (this assumes pod IDs are not guaranteed unique). With this change, I don't think we could delete pod `abc` until 10:10am.

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for your pr @tedyu :)

Do you mind sharing more about the real world impacts of extra locking overhead? I.e. are you experiencing performance issues or other types of resource contention?

My personal opinion is that the code that exists now (with the extra locking overhead) is more readable and easier to conceptualize than the more efficient new code. I think my preference would be to leave the code the way it is, unless the extra locking has a significant penalty.

Just my opinion :) Would love to hear from others!

File: pkg/kubelet/kubelet_pods.go (hunk 1 above)
Comment (tedyu) [reply pkg/kubelet/kubelet_pods.go]: Good question - I myself is curious to know :-)

**Pull request closed without merging**